pkg/pillar/cmd/zedagent/handlentp.go

```diff
@@ -0,0 +1,297 @@
+	infoNTPSources *info.ZInfoNTPSources, dest destinationBitset) {
+
+	// TODO: we don't support LPS
+	if (dest &^ LPSDest) == 0 {
+		return
+	}
+	infoMsg := &info.ZInfoMsg{
+		Ztype: info.ZInfoTypes_ZiNTPSources,
+		DevId: devUUID.String(),
+		InfoContent: &info.ZInfoMsg_NtpSources{
+			NtpSources: infoNTPSources,
+		},
+		AtTimeStamp: ptypes.TimestampNow(),
+	}
+
+	log.Functionf("publishNTPSourcesToDest: sending %v", infoMsg)
+	data, err := proto.Marshal(infoMsg)
+	if err != nil {
+		log.Fatal("publishNTPSourcesToController: proto marshaling error: ", err)
+	}
+	buf := bytes.NewBuffer(data)
+	if buf == nil {
+		log.Fatal("malloc error")
+	}
+	size := int64(proto.Size(infoMsg))
+
+	const bailOnHTTPErr = false
+	const withNetTrace = false
+	key := "ntpsources:" + devUUID.String()
+
+	// Even for the controller destination we can't stall the queue on error,
+	// because this is recurring call, so set @forcePeriodic to true
+	forcePeriodic := true
+	queueInfoToDest(ctx, dest, key, buf, size, bailOnHTTPErr, withNetTrace,
+		forcePeriodic, info.ZInfoTypes_ZiNTPSources)
+}
+
+// createNTPSource() returns `info.NTPSource`. The code is based on the
+// https://github.com/facebook/time/blob/main/cmd/ntpcheck/checker
+func createNTPSource(s *chrony.ReplySourceData,
+	p *chrony.ReplyNTPData,
+	n *chrony.ReplyNTPSourceName) (*info.NTPSource, error) {
+
+	// Clear auth and interleaved flag
+	flash := s.Flags & chrony.NTPFlagsTests
+	// Don't report all flashers if peer is unreachable
+	if flash > 0 {
+		flash ^= chrony.NTPFlagsTests
+	}
+	// all NTP measures are in ms, and chrony reports all in seconds, thus secToMS everywhere
+	ntpSource := info.NTPSource{
+		Authenticated: (s.Flags & chrony.NTPFlagAuthenticated) != 0,
+		Reachable:     s.Reachability == 255, // all 8 attempts
+		Reachability:  uint32(s.Reachability),
+		// We have to advance on 1 due to the UNSPECIFIED enum in protobuf
+		Mode: info.NTPSourceMode(s.Mode + 1),
+		// We have to advance on 1 due to the UNSPECIFIED enum in protobuf
+		State: info.NTPSourceState(s.State + 1),
+		Flags: uint32(flash),
+		// sourceData offset and NTPData offset sign has opposite meaning
+		Offset:  -1 * s.OrigLatestMeas,
+		Poll:    int32(s.Poll),
+		Stratum: uint32(s.Stratum),
+		SrcAddr: s.IPAddr.String(),
+	}
+	if ntpSource.Mode > info.NTPSourceMode_NTP_SOURCE_MODE_REF {
+		ntpSource.Mode = info.NTPSourceMode_NTP_SOURCE_MODE_UNSPECIFIED
+	}
+	if ntpSource.State > info.NTPSourceState_NTP_SOURCE_STATE_OUTLIER {
+		ntpSource.State = info.NTPSourceState_NTP_SOURCE_STATE_UNSPECIFIED
+	}
+
+	// Populate data from NTPData struct
+	if p != nil {
+		refID := chrony.RefidAsHEX(p.RefID)
+		// Only stratum 1 servers can have GPS or something else as string refID
+		if p.Stratum == 1 {
+			refIDStr := chrony.RefidToString(p.RefID)
+			if len(refIDStr) > 0 {
+				refID = refIDStr
+			}
+		}
+		ntpSource.Leap = uint32(p.Leap)
+		ntpSource.Poll = int32(p.Poll)
+		ntpSource.DstAddr = p.LocalAddr.String()
```

Comment:
Should be `SrcAddr`

Reply:
I will take a look deeply, but look what is inside chrony itself:
this is the `RemoteAddr` and `LocalAddr` for `NTPData`: 
https://gitlab.com/chrony/chrony/-/blob/master/cmdmon.c?ref_type=heads#L1237

and here is the `IPAddr` for the `SourceData`:
https://gitlab.com/chrony/chrony/-/blob/master/cmdmon.c?ref_type=heads#L598

gut feeling says, that everything from `NTPData` is from reported from NTP server, 'local address of NTP' is actual a destination for chrony. I will recheck that tomorrow. Anyway thanks.

Reply:
So fields were set from NTP peer perspective: src is the address of NTP and dst is the one who is connected to be synchronized. In our case that seems to be confusing, so I swapped those two. Thanks.